examples/simulation.rs

```diff
@@ -439,7 +439,9 @@ fn main() {
             .build_with_transactions(txs.clone(), rng)
             .expect("instantiate QueueingHoneyBadger");
         let (sq, mut step) = SenderQueue::builder(qhb, peer_ids.into_iter()).build(our_id);
-        assert!(step.extend_with(qhb_step, Message::from).is_empty());
+        assert!(step
+            .extend_with(qhb_step, |fault| fault, Message::from)
+            .is_empty());
```

Comment:
Might be more readable to just assert that both `step` and `qhb_step` are empty. No need to `extend_with`. (Also below.)

Reply:
`Step.extend_with` returns its argument's `output` field, in this case that would be `qhb_step.output`, so it seems that this assertion is equivalent to `assert!(qhb_step.output.is_empty())` not to `assert!(step.is_empty() && qhb_step.is_empty())`

Is that correct or did you mean something else?

Reply:
Ah, you're right, of course, sorry! :+1:

Reply:
I think assertions should be side-effect free! If the `assert!` macro were to be turned off, the code would break!

If this was `debug_assert!`, it would definitely be wrong, I am not 100% sure about the `assert!` macro though - but the fact that I even have to think about/research it make me prefer to have all the actual work done outside the assert.

Reply:
And yes, I realize that this was the case even before this PR, so the question is whether or not we want to change it now in one fell swoop.

Reply:
`step` is defined as `mut` in this piece of code, and `step.extend_with(...)` actually changes it's state, then `step` is returned from the closure. So it seems that in order to preserve the current functionality intact, it would be necessary to call `extend_with` anyway. In such a case we can keep the assert as it is now.
If assertion is changed to something like `assert!(step.is_empty() && qhb_step.is_empty())` the ci indeed fails.

Reply:
So I suggest it's not changed (at least in this PR)

Reply:
I agree, we should assign this to a local `output` variable and _then_ assert that `output.is_empty()`. But I'm fine with postponing this change to another PR, too, if you prefer.